### PR TITLE
XKit 7.9.0

### LINKED ISSUE #192: Investigate post_listener performance improvements
post_listener currently requires each separate extension to implement their own version of `$(".post").not("xwhatever-done")`. There is an opportunity to easily speed this up by refactoring post_listener to be more intelligent. Ideally, post_listeners would receive a list of new posts or be called for each new post.

### LINKED ISSUE #275: Replace uses of alert() with XKit.window.show
Basic browser alerts are ugly, `XKit.window.show` is a prettier way of communicating errors to the user. This could also involve creating a better api for `XKit.window.show`

### LINKED ISSUE #483: Deprecate XKit.console
`XKit.console` is a mess and really isn't all that useful to us for debugging.  It should be deprecated in favour of native `console.log` where absolutely necessary.

### LINKED ISSUE #493: Chrome's GM_listValues looks in the wrong place
as noted by @nightpool in [xkit_patches.js#L228](https://github.com/new-xkit/XKit/blob/master/Extensions/xkit_patches.js#L228), chrome's `GM_listValues()` doesn't work; it looks in `window.localStorage` which is empty unless you have Retags.  This causes the check on line 228 to go awry and the full config dump is empty.

### LINKED ISSUE #1190: xkit-extension.tumblr.com dead, several links not working
Broken links are:
- `bridge.js`
  - [L76](https://github.com/new-xkit/XKit/blob/1375a0e188e5bba3e85e4757506bbd1828f2ded4/Chrome/bridge.js#L76)
  - [L102](https://github.com/new-xkit/XKit/blob/1375a0e188e5bba3e85e4757506bbd1828f2ded4/Chrome/bridge.js#L102)
- `xkit_installer.js`
  - [L43](https://github.com/new-xkit/XKit/blob/1375a0e188e5bba3e85e4757506bbd1828f2ded4/Extensions/xkit_installer.js#L43)
- `xkit_patches.js`
  - [L26](https://github.com/new-xkit/XKit/blob/f16525cc99cb9b82d4400836bf1a1ca70a492577/Extensions/xkit_patches.js#L26)
- `people_notifier.js` (not actual link, just rebranding)
  - [L486](https://github.com/new-xkit/XKit/blob/71585beeed796a9411583f2bf78bff3851711e7c/Extensions/people_notifier.js#L486)
  - [L498](https://github.com/new-xkit/XKit/blob/71585beeed796a9411583f2bf78bff3851711e7c/Extensions/people_notifier.js#L498)
- `old_notifications.js` (same as above)
  - [L4](https://github.com/new-xkit/XKit/blob/1375a0e188e5bba3e85e4757506bbd1828f2ded4/Extensions/old_notifications.js#L4)
- `servant.js` (same as above)
  - [L122](https://github.com/new-xkit/XKit/blob/1375a0e188e5bba3e85e4757506bbd1828f2ded4/Extensions/servant.js#L122)
  - [L150](https://github.com/new-xkit/XKit/blob/1375a0e188e5bba3e85e4757506bbd1828f2ded4/Extensions/servant.js#L150)

### LINKED ISSUE #1478: troubleshooting thing links to old Xkit troubleshooter
Saw new version of noscript, thought I'd see if new xkit could reach update servers, it still can't, but now it offers me a troubleshooting link that's definitely for old xkit. (No reference to versions past 7.4, etc.)

It'd maybe help just to have, say, a list of domains that need to be whitelisted, or something? Unless it just plain can't be run at all ever with noscript until that updates. In which case it might be nice to have a version which, say, doesn't constantly spam the error messages, because as is, it's not actually practical to try to use tumblr at all with both new xkit and noscript...

### LINKED ISSUE #1537: both Configuration Exports entirely broken
`XKit.tools.make_gist()` is now completely useless, following [the removal of anonymous gists](https://blog.github.com/2018-02-18-deprecation-notice-removing-anonymous-gist-creation/) last month. Exporting a simple text file could probably work instead since both GitHub and Discord support file uploads.

### LINKED ISSUE #1557: Firefox Multi-Account Containers prevents One-Click Reblog from working
Fairly easy to reproduce - first, log in on a regular tab. Then, log in to a seperate account in a container tab. One-click reblogs works on the initial tab, however, in the container tab, reblogs won't work. This is due to the extension detecting the uncontainerised account instead of the containerised one.

The plugin works fine when using seperate Firefox profiles, but the extension to easily manage those was deprecated with the release of Quantum and most users would be using MAC now, due to it being pushed by Mozilla as a way to compartmentalise and improve user privacy/protection. 

Link to extension: https://addons.mozilla.org/en-US/firefox/addon/multi-account-containers/

### PR DESCRIPTION
**Pull request by AprilSylph**
- [x] refactor XKit Patches
- [x] run at document start
- [x] remove XKit.console 
- [x] post listener improvements/rewrite
- [x] built-in troubleshooting upon installation failure
- [x] remove `alert()` overusage
- [x] move any vital information to a `XKit.window` call
- [x] remove `XKit.tools.make_gist` in favour of a `make_file()` tool
- [x] cleanup
    - [x] remove references to `XKit.page.peepr`
    - [x] remove non-link references to xkit-extension
    - [x] create empty 7.9.0 patches function
    - [x] version bumps for extensions

---
Refactor XKit Patches
---
puts most of xkit patches into xkit.js because quite frankly half of the XKit object being in xkit patches is silly. this isn't noticeably more efficient on its own but does make code lookup a lot easier

most of XKit Patches' code (i.e. any function defining or redefining) is now executed based on the XKit version running it; the additions to XKit Patches made in #1511 is a great example of a patch which  becomes useless in a certain XKit version, and has been treated accordingly, not running on v7.8.2 or above. All patches that 7.8.2 relies on will still run on 7.8.2 but have been added to 7.9.0 directly and thus will not run there.

Points of contention:
- This changes the release process. Each new version of XKit must have the previous version's patches shuffled into it, and must be given its own (empty) patches function.
- This deprecates the XKit Patches CSS file. What would be changes to it would now be put in the JS file (in the form of `XKit.tools.add_css()`) and then added to xkit.css for the next release.

diff best viewed in no whitespace-only changes mode

---
Prepare at document start
---
Runs every content script at the start of the document instead of the end - every library we use and the entire XKit object is now readied as soon as the page starts loading, and `XKit.init_extension` is bound to `$(document).ready` - the user-facing result is the apparent boot time being reduced to just XKit Main's run time. On my computer, this is the same time Tumblr takes to load Recommended Blogs in the sidebar or the Activity sparkline - blink, and you miss it.

This also simplifies the logic for `XKit.page` and `XKit.init`, and removes `XKit.page.peepr` since it's literally never true now.

---
Deprecate XKit.console
---
Entirely removes the XKit Console and replaces each call for `XKit.console.add` with the appropriate `console` command. Resolves #483.

---
Post Listener rewrite
---
Uses a MutationObserver for the post listener, and makes the callback storage object-based instead of two arrays - this allows for an extension to add multiple functions under the same name, and remove them all at once in its `destroy()` logic.
The MutationObserver's benefits are a lot more noticeable; the post listener is now completely dormant until something happens, and activates as soon as it verifies new posts exist (as opposed to waiting up to 1s to do anything).
Resolves #192 since `XKit.interface.get_posts()` is a thing now and it's wayyy too much work to integrate directly into the post listener.

---
Installation failure troubleshooting tool
---
Provides the same troubleshooting tool upon installation failure as XKit Updates (#1532), with the text altered to fit the context. The script blocker section (part 2) receives a minor style update in both cases, with the window now being wide and `new-xkit.github.io` being inline. Resolves #1478.

---
Remove most `alert()` usage
---
Simply finds alternatives to using `alert` calls, whether it be `XKit.window.show` calls or highlighting input boxes in red when they require a value. Resolves #275.

---
Use `Nx_XHR` in `XKit.interface.kitty.get()`
---
Resolves #1557 and contains changes to `Nx_XHR` to make headers available to `onload` logic (in `response.headers`).

---
Remove old XKit links
---
Moves any vital information to an `XKit.window` call, most notably the xkit-extension post about corrupt profiles, Timestamps syntax formatting, and the Servant tutorial. Resolves #1190, replaces and closes #1191.

---
Fix configuration exports
---
- Replaces `XKit.tools.make_gist` with a handy dandy `make_file` tool
    - resolves #1537 
- Fixes the Full Configuration Export by rewriting `XKit.tools.dump_config` to not use `GM_listValues()`
    - somewhat integrates, closes #711 
    - resolves #493

### CODE REVIEW

**Comment by AprilSylph** [comment]
Okay so, there IS actually a use for this (sort of).
If we run all the main scripts at document start, then wait for the document to be ready before calling XKit Main, we get a noticeable decrease in apparent boot time. I haven't tested this without the patches refactor so it *could* just be all the libraries we use, but I still think defining as much as possible pre-initialisation is the way to go.
If there are no concerns about the changes to the release process this PR suggests, I could add that commit here too as to bundle it all into one release.

**Comment by AprilSylph** [comment]
just removed the empty 7.8.3 function, seeing as there won't be a release 7.8.3 per discussion in #1542

File: Extensions/xkit_patches.css [hunk 1]
```diff
@@ -1,3 +1,9 @@
+/* ⚠️ DEPRECIATED ⚠️
```
Comment (ThePsionic) [inline Extensions/xkit_patches.css]: Deprecated?

**Comment by AprilSylph** [comment]
@hobinjk thoughts re: the release process? if it's okay, I'd like to turn this into a 7.9.0 PR since there's a few other things I'd like to do with xkit.js

**Comment by hobinjk** [comment]
All of this SGTM as long as there isn't a significant load time regression, thanks for the in-depth explanation!

**Comment by AprilSylph** [comment]
updated the initial comment with everything that's going on and will continue to update it with each (notable) commit

**Comment by AprilSylph** [comment]
terrifying

**Comment by AprilSylph** [comment]
aaand I'm pretty sure I'm done! Initial comment edited to document everything I did, and tested on all three major browsers (although I weirdly couldn't connect to my dev server on safari, But I Didn't Touch The Safari Request Method So It's Not My Fault And It Should Be Fine:tm:)

**IMPORTANT** - this release needs a *long* pre-release period for extension updates to propagate. Since I've finally ripped XKit.console out entirely, anyone who gets XKit 7.9.0 before the extension updates will have to dance around versions or just bite the bullet and redo their settings.
(When this gets pulled, I'll make a blog post both explaining the deal with that and advertising the pre-release)

**Comment by ThePsionic** [comment]
Might be worth mentioning XCloud (local) export as a user-generated backup

File: Extensions/blacklist.js [hunk 2]
```diff
@@ -415,30 +435,43 @@ XKit.extensions.blacklist = new Object({
...
+						$m_to_add
+							.removeAttr("style")
+							.attr("placeholder", "Enter a word here.")
+							.off("click");
+					});
+			}
 
 			if (m_to_add === "" || $.trim(m_to_add) === "") {
-				XKit.window.close();
+				complain("Not even XKit can save you from ALL posts.");
```
Comment (hobinjk) [inline Extensions/blacklist.js]: yes good

File: Extensions/one_click_postage.js [hunk 3]
```diff
@@ -1498,7 +1498,7 @@ XKit.extensions.one_click_postage = new Object({
 		try {
 			limit = parseInt(limit_count.substring(1));
 		} catch (e) {
-			//alert("NO");
```
Comment (hobinjk) [inline Extensions/one_click_postage.js]: same

**Pull request merged**